contracts/Parameterizer.sol

```diff
@@ -1,96 +1,323 @@
+  }
+
+  /**
+  @notice Determines whether a proposal exists for the provided proposal ID
+  @param _propID The proposal ID whose existance is to be determined
+  */
+  function propExists(bytes32 _propID) constant public returns (bool) {
+    return proposalMap[_propID].processBy > 0;
+  }
+
+  /**
+  @notice Determines whether the provided proposal ID has a challenge which can be resolved
+  @param _propID The proposal ID whose challenge to inspect
+  */
+  function challengeCanBeResolved(bytes32 _propID) constant public returns (bool) {
+    ParamProposal memory prop = proposalMap[_propID];
+    Challenge memory challenge = challengeMap[prop.challengeID];
+
+    return (prop.challengeID > 0 && challenge.resolved == false &&
+            voting.pollEnded(prop.challengeID));
+  }
+
+  /**
+  @notice Determines the number of tokens to awarded to the winning party in a challenge
+  @param _challengeID The challengeID to determine a reward for
+  */
+  function determineReward(uint _challengeID) public constant returns (uint) {
+    require(!challengeMap[_challengeID].resolved && voting.pollEnded(_challengeID));
+
+    if(voting.getTotalNumberOfTokensForWinningOption(_challengeID) == 0) {
+      // Edge case, nobody voted, give all tokens to the winner.
+      return 2 * challengeMap[_challengeID].stake;
+    }
+    
+    return (2 * challengeMap[_challengeID].stake) - challengeMap[_challengeID].rewardPool;
+  }
+
+  /**
+  @notice gets the parameter keyed by the provided name value from the params mapping
+  @param _name the key whose value is to be determined
+  */
+  function get(string _name) public constant returns (uint value) {
+    return params[keccak256(_name)];
+  }
+
+  // ----------------
+  // PRIVATE FUNCTIONS:
+  // ----------------
+
+  /**
+  @dev sets the param keted by the provided name to the provided value
+  @param _name the name of the param to be set
+  @param _value the value to set the param to be set
+  */
+  function set(string _name, uint _value) private {
+    params[keccak256(_name)] = _value;
+  }
+
+  /**
+  @dev resolves a challenge for the provided _propID. It must be checked in advance whether the _propID has a challenge on it
+  @param _propID the proposal ID whose challenge is to be resolved.
+  */
+  function resolveChallenge(bytes32 _propID) private {
+    ParamProposal memory prop = proposalMap[_propID];
+
+    // winner gets back their full staked deposit, and dispensationPct*loser's stake
+    uint reward = determineReward(prop.challengeID);
+
+    if (voting.isPassed(prop.challengeID)) { // The challenge failed
+      if(prop.processBy > now) {
+        set(prop.name, prop.value);
+      }
+      require(token.transfer(prop.owner, reward));
+    } 
+    else { // The challenge succeeded
+      require(token.transfer(challengeMap[prop.challengeID].challenger, reward));
+    }
+
+    // set flag on challenge being processed
+    challengeMap[prop.challengeID].resolved = true;
```

Comment:
as a design patter, should the state transition be done before the transfer (even though re-entrancy may not possible here).